src/savegame_json.cpp

```diff
@@ -1528,22 +1530,71 @@ void item::io( Archive& archive )
-            item_tags.erase( "DOUBLE_AMMO" );
-            emplace_back( "battery_compartment" );
-
-        } else if( has_flag( "USE_UPS" ) ) {
-            item_tags.erase( "USE_UPS" );
-            emplace_back( "battery_ups" );
-
-        } else if( has_flag( "DOUBLE_REACTOR" ) ) {
-            item_tags.erase( "DOUBLE_REACTOR" );
-            emplace_back( "double_plutonium_core" );
+    if( is_tool() || is_toolmod() ) {
+        migrate_toolmod( *this );
+    }
+}
+
+static void migrate_toolmod( item &i )
+{
+    // Convert legacy flags on tools to contained toolmods
+    if( i.is_tool() ) {
+        if( i.item_tags.count( "ATOMIC_AMMO" ) ) {
+            i.item_tags.erase( "ATOMIC_AMMO" );
+            i.item_tags.erase( "NO_UNLOAD" );
+            i.item_tags.erase( "RADIOACTIVE" );
+            i.item_tags.erase( "LEAK_DAM" );
+            i.emplace_back( "battery_atomic" );
+
+        } else if( i.item_tags.count( "DOUBLE_REACTOR" ) ) {
+            i.item_tags.erase( "DOUBLE_REACTOR" );
+            i.item_tags.erase( "DOUBLE_AMMO" );
+            i.emplace_back( "double_plutonium_core" );
+
+        } else if( i.item_tags.count( "DOUBLE_AMMO" ) ) {
+            i.item_tags.erase( "DOUBLE_AMMO" );
+            i.emplace_back( "battery_compartment" );
+
+        } else if( i.item_tags.count( "USE_UPS" ) ) {
+            i.item_tags.erase( "USE_UPS" );
+            i.item_tags.erase( "NO_RELOAD" );
+            i.item_tags.erase( "NO_UNLOAD" );
+            i.emplace_back( "battery_ups" );
+
+        }
+    }
+
+    // Fix fallout from #18797, which exponentially duplicates migrated toolmods
+    if( i.is_toolmod() ) {
+        // duplication would add an extra toolmod inside each toolmod on load;
+        // delete the nested copies
+        if( i.typeId() == "battery_atomic" || i.typeId() == "battery_compartment" ||
+            i.typeId() == "battery_ups" || i.typeId() == "double_plutonium_core" ) {
+            // Be conservative and only delete nested mods of the same type
+            i.contents.remove_if( [&]( const item &cont ) {
+                    return cont.typeId() == i.typeId();
+                        } );
+        }
+    }
+
+    if( i.is_tool() ) {
+        // duplication would add an extra toolmod inside each tool on load;
+        // delete the duplicates so there is only one copy of each toolmod
+        int n_atomic = 0,
+            n_compartment = 0,
+            n_ups = 0,
+            n_plutonium = 0;
+
+        // not safe to use remove_if with a stateful predicate
+        for( auto it = i.contents.begin(); it != i.contents.end(); ) {
+            if( ( it->typeId() == "battery_atomic" && ++n_atomic > 1 ) ||
+                ( it->typeId() == "battery_compartment" && ++n_compartment > 1 ) ||
+                ( it->typeId() == "battery_ups" && ++n_ups > 1 ) ||
+                ( it->typeId() == "double_plutonium_core" && ++n_plutonium > 1 ) ) {
+                i.contents.erase( it++ );
```

Comment:
Either postfix increment or assign the result, one or the other, which is preferred?